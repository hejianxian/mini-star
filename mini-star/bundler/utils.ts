import fs from 'fs';
import path from 'path';

/**
 * Find plugin dir names
 */
export function getPluginDirs(): string[] {
  const list = fs.readdirSync(path.resolve(process.cwd(), './plugins/'));

  const plugins = list.filter((item) =>
    fs.statSync(path.resolve(process.cwd(), './plugins/', item)).isDirectory()
  );

  return plugins;
}
