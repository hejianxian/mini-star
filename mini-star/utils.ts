/**
 * Create blank ModuleLoader
 * @returns A new ModuleLoader
 */
export function createNewModuleLoader(): ministar.ModuleLoader {
  return {
    resolves: [],
    status: 'new',
    entryFn: null,
    ins: null,
  };
}

/**
 * Set module loader status to `loaded`
 */
export function setModuleLoaderLoaded(
  moduleLoader: ministar.ModuleLoader,
  moduleExport: {} | null
) {
  moduleLoader.status = 'loaded';
  moduleLoader.ins = moduleExport;
  if (Array.isArray(moduleLoader.resolves)) {
    moduleLoader.resolves.forEach((_resolve) => {
      _resolve(moduleExport);
    });
  }
  moduleLoader.resolves = [];
}

function trimDotOfPath(parent: string, self: string) {
  if (!self.startsWith('.')) {
    return self;
  }
  const parentList = parent.split('/');
  const selfList = self.split('/');
  parentList.pop();
  const selfRet: string[] = [];
  selfList.forEach((item) => {
    if (item === '..') {
      parentList.pop();
    }
    if (item !== '..' && item !== '.') {
      selfRet.push(item);
    }
  });
  return parentList.concat(selfRet).join('/');
}

/**
 * A function how to process module path to real path
 */
export function processModulePath(moduleName: string, path: string): string {
  if (path.startsWith('@plugin/')) {
    const name = path.split('/')[1];
    return `${name}/index.js`;
  }

  if (path.endsWith('.js') || path.startsWith('./')) {
    return trimDotOfPath(
      moduleName,
      !path.endsWith('.js') ? `${path}.js` : path
    );
  }

  return path;
}
