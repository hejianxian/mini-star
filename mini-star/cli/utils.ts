import path from 'path';
import { config } from '../bundler/config';

export function getPluginPackagePath(pluginName: string) {
  return path.resolve(
    config.pluginRoot,
    './plugins/',
    pluginName,
    './package.json'
  );
}

/**
 * Return HH:mm
 */
export function getShortTimeStr(): string {
  const now = new Date();

  return `${String(now.getHours()).padStart(2, '0')}:${String(
    now.getMinutes()
  ).padStart(2, '0')}`;
}
