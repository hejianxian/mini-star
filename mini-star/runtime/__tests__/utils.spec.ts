import { createNewModuleLoader, processModulePath } from '../utils';

test('createNewModuleLoader', () => {
  expect(createNewModuleLoader()).toMatchSnapshot();
});

describe('processModulePath', () => {
  test.each([
    ['require', 'require'],
    ['exports', 'exports'],
    ['model', 'model'],
    ['model.js', 'model.js'],
    ['./model', '@plugins/demo/model.js'],
    ['./model.js', '@plugins/demo/model.js'],
    ['@plugins/common', '@plugins/common'],
  ])('%s => %s', (input, output) => {
    expect(processModulePath('@plugins/demo', input)).toBe(output);
  });

  test('deep path', ()=> {
    expect(processModulePath('@plugins/demo/index-353b8484.js', './index-4a4caf9c.js')).toBe('@plugins/demo/index-4a4caf9c.js')
  })
});
