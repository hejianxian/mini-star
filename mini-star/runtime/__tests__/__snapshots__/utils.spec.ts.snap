// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNewModuleLoader 1`] = `
Object {
  "entryFn": null,
  "ins": null,
  "resolves": Array [],
  "status": "new",
}
`;
